test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/reflect/ReflectArrayTest.scala

```diff
@@ -0,0 +1,68 @@
+/*                     __                                               *\
+**     ________ ___   / /  ___      __ ____  Scala.js Test Suite        **
+**    / __/ __// _ | / /  / _ | __ / // __/  (c) 2013, LAMP/EPFL        **
+**  __\ \/ /__/ __ |/ /__/ __ |/_// /_\ \    http://scala-js.org/       **
+** /____/\___/_/ |_/____/_/ | |__/ /____/                               **
+**                          |/____/                                     **
+\*                                                                      */
+package org.scalajs.testsuite.javalib.lang.reflect
+
+import scala.runtime.BoxedUnit
+
+import org.junit.Test
+import org.junit.Assert._
+
+class ReflectArrayTest extends ReflectArrayTestCommon {
+
+  @Test def `newInstance`(): Unit = {
+    test(classOf[Int], classOf[Array[Int]], 0)
+    test(classOf[Char], classOf[Array[Char]], '\0')
+    test(classOf[Long], classOf[Array[Long]], 0L)
+    test(classOf[Boolean], classOf[Array[Boolean]], false)
+
+    test(classOf[BoxedUnit], classOf[Array[Unit]], null) // yes, null
+
+    test(classOf[Object], classOf[Array[Object]], null)
+    test(classOf[String], classOf[Array[String]], null)
+
+    test(classOf[java.lang.Integer], classOf[Array[java.lang.Integer]], null)
+    test(classOf[java.lang.Long], classOf[Array[java.lang.Long]], null)
+
+    test(classOf[Array[Object]], classOf[Array[Array[Object]]], null)
+    test(classOf[Array[Int]], classOf[Array[Array[Int]]], null)
+    test(classOf[Array[String]], classOf[Array[Array[String]]], null)
+
+  }
+}
+
+trait ReflectArrayTestCommon {
+
+  import java.lang.reflect.Array.newInstance
+
+  @inline
+  private def testBase(clazz: Class[_], length: Int, expectedClazz: Class[_],
+      sampleElem: Any): Unit = {
+    val array = newInstance(clazz, length).asInstanceOf[Array[_]]
+    assertTrue(array.getClass == expectedClazz)
+    assertTrue(array.getClass.isArray)
+    assertEquals(length, array.length)
+    for (i <- 0 until array.length)
+      assertTrue(array(i) == sampleElem)
+  }
+
+  @noinline
+  private def testNoInline(clazz: Class[_], length: Int, expectedClazz: Class[_],
+      sampleElem: Any): Unit = {
+    testBase(clazz, length, expectedClazz, sampleElem)
+  }
+
+  @inline
+  def test(clazz: Class[_], expectedClazz: Class[_],
+      sampleElem: Any): Unit = {
```

Comment:
This one is more involved, so maybe copy-pasting it is not such a good idea. But at least the name of this method and its helpers should be `testNewInstance`.